src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs

```diff
@@ -2148,31 +2148,49 @@ private string[] GetNativeDigits()
                 return result;
             }
 
-            // Try to check if the digits are all ASCII so we can avoid the array allocation and use the static array NumberFormatInfo.s_asciiDigits instead.
-            // If we have non-ASCII digits, we should exit the loop very quickly.
-            int i = 0;
-            while (i < NumberFormatInfo.s_asciiDigits.Length)
-            {
-                if (digits[i] != NumberFormatInfo.s_asciiDigits[i][0])
-                {
-                    break;
-                }
-                i++;
-            }
+            // In ICU we separate the digits with the '\uFFFF' character
 
-            if (i >= NumberFormatInfo.s_asciiDigits.Length)
+            if (digits.StartsWith("0\uFFFF1\uFFFF2\uFFFF3\uFFFF4\uFFFF5\uFFFF6\uFFFF7\uFFFF8\uFFFF9\uFFFF", StringComparison.Ordinal) ||  // ICU common cases
+                digits.StartsWith("0123456789", StringComparison.Ordinal))  // NLS common cases
             {
                 return result;
             }
 
-            // we have non-ASCII digits
+            // Non-ASCII digits
+
+            // Check if values coming from ICU separated by 0xFFFF
+            int ffffPos = digits.IndexOf('\uFFFF');
+
             result = new string[10];
-            for (i = 0; i < result.Length; i++)
+            if (ffffPos < 0) // NLS case
             {
-                result[i] = char.ToString(digits[i]);
+                for (int i = 0; i < result.Length; i++)
+                {
+                    result[i] = char.ToString(digits[i]);
+                }
+
+                return result;
             }
 
-            return result;
+            // ICU case
+
+            int start = 0;
+            int index = 0;
+
+            do
+            {
+                result[index++] = digits.Substring(start, ffffPos - start);
+                start = ++ffffPos;
+                while ((uint)ffffPos < (uint)digits.Length && digits[ffffPos] != '\uFFFF')
+                {
+                    ffffPos++;
+                }
+
+            } while (ffffPos < digits.Length && index < 10);
+
+            Debug.Assert(index >= 10, $"Couldn't read native digits for '{_sWindowsName}' successfully.");
+
+            return index < 10 ? NumberFormatInfo.s_asciiDigits : result;
```

Comment:
We just asserted that index >= 10.  Is there any situation where we expect index < 10?

Reply:
We can get less than 10 if we get bad data from the ICU. This is unlikely and that is why I have it as assert here.

Reply:
I'm just struggling with what kind of bad data we make accomodations for.  We could, for example, get bad enough data that we overflow a buffer, but we're not protecting against that.

Reply:
In the native side, we fill the buffer with the data. Every native digit data is split with `\uFFFF`. We have a good check there guarantee we'll not overflow the buffer. You had the comment https://github.com/dotnet/runtime/pull/71045#discussion_r903234698 which at the end you have suggesting adding assert. Such assert would be useful only if we get ICU data which makes the buffer not enough to hold all the data. If this happens (which is very unlikely), we'll not have all the data in the buffer and we'll not end up having 10 native digits data in the buffer. This is the assertion I have on the managed side. 

We are not going to have any buffer overflow in any case.

Reply:
If `unum_getSymbol` were to return a really large symbol length, there's no possibility we might overflow the value buffer?

Reply:
Ah, I see, ok.